### PR TITLE
Implemented Course view

### PR DESCRIPTION
**Pull request by chizuo**
This feature implements the `Course` controller & view, respectively named [`CourseController`](https://github.com/chizuo/COMP586-Project/blob/main/Controllers/CourseController.cs) and [`Index.cshtml`](https://github.com/chizuo/COMP586-Project/blob/main/Views/Course/Index.cshtml).

The `CourseController`  only has one route, `index(string course)` which currently is permitted to be nullable for testing purposes. But, in the future, it will require a query string in the form of `https://localhost:7150/Course/index?course=string` where the string is `Course.Code`.  The `Course` view will require a `Course` object as its Model and will unbox the `Course.PreReq` and `Course.CoReq` to create links that will route to the same endpoint with a query string of the course found in each respective `List<Course`, that is the property `Couse.PreReq` and `Course.CoReq`. Once the database has been implemented with data to test, the `index(string course)` method will use the string as the unique ID to query the database for the course information. The `string course` represents the unique ID of the course entity in the Course Table of our database.

An additional feature added is the [`Department`](https://github.com/chizuo/COMP586-Project/pull/30/files#diff-83deb0674b8fcf82feb6d7ddbf1144e329b0ceb1803872e92259b2e8463fbd1a) class. This class was created to mimic the implementation we often see universities do. Though, for the scope of this project, it was really used to help make sense of how to create a `course.Code` since part of the code was based from the department's choice of acronym (e.g. Computer Science = COMP, Electrical and Computer Engineer = ECE) and the concatenation of the `Course.number`. I didn't think it made sense to require the creator of the `Course` to provide a `course.Code` in the constructor while also allowing free reign on choosing what the `Course.Code` would be, as it leads to possible shift in convention. Since we intend to use the `Course.Code` as the unique i.d. of the `Course`  entity in the Course Table in our database, it made more sense to enforce the convention through the application to prevent any possible collisions on the unique IDs that are not automatically generated by the database.

The `Department` class is comprised of:
```
protected string name; //representing the department name, like "Computer Science"
public string Name; //the property  that gives you access to read the department name, e.g. department.Name
protected string code; //representing the acronym of the department, e.g. "Computer Science" is "COMP"
public string Code; //the property that gives you access to read the acronym of the department.
protected Professor professor; //professor object that represents the chair of the department
public Professor Chair; //the property that returns or replaces the professor that is the chair of the department
```

The [`Course`](https://github.com/chizuo/COMP586-Project/blob/main/Models/Course.cs) class has been updated with the following changes:

- Removed getter methods and replaced them with properties to simplify the code and adhere more towards the C# convention of getters and setters for protected / private fields.
- changed the field `string courseName` to `string subject`.
- changed `string dept` to `Department department`, which is now an object of class `Department`
- created a property `public string Name` that returns the concat of the `department.Code` with the course number and subject for the purpose of simplifying the view portion of the application. E.g.`course.Name` would return something like "COMP586 : Object Oriented Design"/
- created a property 'public string Code' that returns the concat of the `department.Code` with the course number. This was done to help simplify the query string creation of any `View()` in the application that intends to route to the Course end point by using the Razor annotation `<a href="./Course/index?course=@course.Code">@course.Code</a>`
- created a property `public List<Course> CoReq` and `public List<Course> PreReq` for the purpose of creating a simple iterable object for the `View()` that intends to unbox and create clickable links for the course route by using the `Course` object property.
- created the properties `public HashSet<Course> PreRequisites` and `public HashSet<Course> CoRequisites` as the setters for a `Course` object's `coRequisites` and `preRequisites`
- created properties and fields related to lab. 
- `isLab` is used to identify that the course is a lab component, not lecture. Thus, the `course.Code` will append "L" at the end of the `course.Code`, e.g. "COMP586L"
- `Course lab` is the field that holds the lab component `Course` object. I'm deciding if this makes sense or if it makes more sense to just hold the `course.Code` of the lab component since it is always the `course.Code` with L appended to it.

### CODE REVIEW

File: Views/Course/Index.cshtml [hunk 1]
```diff
@@ -5,5 +5,17 @@
...
 
 <div id="course">
-    <h1>Course @Model.dept @Model.number : @Model.name (@Model.units)</h1>
-</div>
\ No newline at end of file
+    <h2 id="course-header"><span>Course:</span> @Model.Name (@Model.Units)</h2>
+    <hr>
+    <div id="course-card">
+        <div id="contents">
+            <p><span>Corequisites:</span> @foreach(var course in Model.PreReq){
```
Comment (harshleen8) [inline Views/Course/Index.cshtml:12]: Here, I think we should be iterating it with Model.CoReq isn't? foreach(var course in Model.CoReq)

File: Views/Course/Index.cshtml [hunk 2]
```diff
@@ -5,5 +5,17 @@
...
+                <a href="./index?course=@course.Code">@course.Code</a>
+            }</p>
+            <p><span>Prerequisites:</span> @foreach(var course in Model.CoReq){
```
Comment (harshleen8) [inline Views/Course/Index.cshtml:15]: same thing here.

File: Views/Course/Index.cshtml (hunk 1 above)
Comment (chizuo) [reply Views/Course/Index.cshtml]: this has been [addressed](https://github.com/chizuo/COMP586-Project/pull/30/commits/f488195b9bcbe8f76e950d16910ed928832a454e). Good catch!

**Comment by Brandon-CSUN** [comment]
Code looks greats! Really good in-depth description about the changes and additional feature that was added.

**Comment by Jacob-Asaad** [comment]
Looks good to me, an especially good description of the changes made to the existing code. I have a question that I'll address in the discord.

**Pull request merged**